stl/inc/memory

```diff
@@ -2416,26 +2437,283 @@ bool atomic_compare_exchange_weak_explicit(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty
+_CXX20_DEPRECATE_OLD_SHARED_PTR_ATOMIC_SUPPORT bool atomic_compare_exchange_strong_explicit(shared_ptr<_Ty>* _Ptr,
+    shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value, memory_order, memory_order) { // atomically compare and exchange
     return _STD atomic_compare_exchange_weak_explicit(
         _Ptr, _Exp, _STD move(_Value), memory_order_seq_cst, memory_order_seq_cst);
 }
 
 template <class _Ty>
-bool atomic_compare_exchange_strong(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
+_CXX20_DEPRECATE_OLD_SHARED_PTR_ATOMIC_SUPPORT bool atomic_compare_exchange_strong(
+    shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
     // atomically compare and exchange
     return _STD atomic_compare_exchange_strong_explicit(
         _Ptr, _Exp, _STD move(_Value), memory_order_seq_cst, memory_order_seq_cst);
 }
 
+#if _HAS_CXX20 && !defined(_M_CEE_PURE)
+template <class _Ty>
+struct atomic<shared_ptr<_Ty>> : _Atomic_storage<shared_ptr<_Ty>> {
+private:
+    using _Base = _Atomic_storage<shared_ptr<_Ty>>;
+
+public:
+    using value_type = shared_ptr<_Ty>;
+
+#if 1 // TRANSITION, ABI
+    static constexpr bool is_always_lock_free =
+        sizeof(shared_ptr<_Ty>) <= 8 && (sizeof(shared_ptr<_Ty>) & sizeof(shared_ptr<_Ty>) - 1) == 0;
+
+    _NODISCARD bool is_lock_free() const noexcept {
+        return is_always_lock_free;
+    }
+#else // ^^^ don't break ABI / break ABI vvv
+#if _ATOMIC_HAS_DCAS
+    static constexpr bool is_always_lock_free = sizeof(shared_ptr<_Ty>) <= 2 * sizeof(void*);
+#else // ^^^ _ATOMIC_HAS_DCAS / !_ATOMIC_HAS_DCAS vvv
+    static constexpr bool is_always_lock_free = sizeof(shared_ptr<_Ty>) <= sizeof(void*);
+#endif // _ATOMIC_HAS_DCAS
+
+    _NODISCARD bool is_lock_free() const noexcept {
+#if _ATOMIC_HAS_DCAS
+        return is_always_lock_free;
+#else // ^^^ _ATOMIC_HAS_DCAS / !_ATOMIC_HAS_DCAS vvv
+        return is_always_lock_free || (sizeof(shared_ptr<_Ty>) <= 2 * sizeof(void*) && __std_atomic_has_cmpxchg16b());
+#endif // _ATOMIC_HAS_DCAS
+    }
+#endif // TRANSITION, ABI
+
+    void store(shared_ptr<_Ty> _Value) noexcept {
+        shared_ptr<_Ty> _Old = exchange(_STD move(_Value));
+    }
+
+    void store(shared_ptr<_Ty> _Value, const memory_order _Order) noexcept {
+        shared_ptr<_Ty> _Old = exchange(_STD move(_Value), _Order);
+    }
+
+    _NODISCARD shared_ptr<_Ty> load() const noexcept {
+        shared_ptr<_Ty> _Result = _Base::load();
+        _Result._Incref();
+        return _Result;
+    }
+
+    _NODISCARD shared_ptr<_Ty> load(const memory_order _Order) const noexcept {
+        shared_ptr<_Ty> _Result = _Base::load(_Order);
+        _Result._Incref();
+        return _Result;
+    }
+
+    operator shared_ptr<_Ty>() const noexcept {
+        return load();
+    }
+
+    _NODISCARD shared_ptr<_Ty> exchange(shared_ptr<_Ty> _Value) noexcept {
+        _Value._Incref();
+        return _Base::exchange(_STD move(_Value));

```

Comment:
for two atomic<shared_ptr>s A and B and two threads T1 and T2 if:
T1:
A = B.exchange(A);

T2:
B = A.echange(B);

or similar are those exchanges required to be ordered strictly one after another? Presumably this is supposed to be the sequencial consistency version.

If how does this function prevent incorrect refcounts if another exchange gets ordered between lines 2519 and 2520

Reply:
Here are 3 main scenarios you may have concerns about (they are written in pseudocode). Please correct me if I missed or forgot about something. They all seem to be OK.

`A` and `B` are 2 atomic<shared_ptr>s and T1 and T2 are 2 threads exactly as in your comment. `a` and `b` are 2 refcounts with `a` starting as a refcount of `A` and `b` starting as refcount of `B`. ---> arrows show a consequence of a line. `{...}` brackets show which refcount is held by the object.

### SCENARIO 1 - whole operation happens between these 2 lines

```
// at the beginning a=1 and b=1
~~~~start at thread T1:

T1_Value{} = A.load(); ---> T1_Value{a} (a++) a=2 b=1
// a is held by A na T1_Value


B.exchange(T1_Value):
{
	T1_Value.incref(); ---> (a++) a=3 b=1

~~~~thread switched to T2:

T2_Value{} = B.load(); ---> T2_Value{b} (b++) a=3 b=2
// b is held by B and T2_Value

A.exchange(T2_Value):
{
	T2_Value.incref(); ---> (b++) a=3 b=3
	
	T2_Result = A._Base::exchange(T2_Value); ---> A{b} T2_Result{a}
	// a is held by T2_Result and T1_Value
	// b is held by A, B and T2_Value
}

destroy T2_Value; ---> (b--) a=3 b=2
// b is held by A and B

B.store(T2_Result):
{
	T2_Old = B.exchange(T2_Result);
	{
		T2_Result.incref(); ---> (a++) a=4 b=2
		
		T2_Old = B._Base::exchange(T2_Result); ---> B{a} T2_Old{b}
		// a is held by B, T2_Result and T1_Value
		// b is held by A and T2_Old
	}
	destroy T2_Old; ---> (b--) a=4 b=1
	// b is held by A
}

destroy T2_Result; ---> (a--) a=3 b=1
// a is held by B and T1_Value

~~~~thread switched to T1:

	T1_Result = B._Base::exchange(T1_Value); ---> B{a} (no change in B) T1_Result{a}
	// a is held by B, T1_Value and T1_Result
}

destroy T1_Value; ---> (a--) a=2 b=1
// a is held by B and T1_Result

A.store(T1_Result):
{
	T1_Old = A.exchange(T1_Result):
	{
		T1_Result.incref(); ---> (a++) a=3 b=1

		T1_Old = A._Base::exchange(T1_Result); ---> A{a} T1_Old{b}
		// a is held by A, B and T1_Result
		// b is held by T1_Old
	}
	destroy T1_Old; ---> (b--) a=3 b=0
	// b is destroyed
}

destroy T1_Result; ---> (a--) a=2 b=0
// a is held by A and B

finaly both A and B hold a, a=2, b reached 0 and got destroyed - everything went well
```

### SCENARIO 2 - only exchange (and load() but it shouldn't matter) happens between these 2 lines

```
// at the beginning a=1 and b=1
~~~~start at thread T1:

T1_Value{} = A.load(); ---> T1_Value{a} (a++) a=2 b=1
// a is held by A and T1_Value

B.exchange(T1_Value):
{
	T1_Value.incref(); ---> (a++) a=3 b=1

~~~~thread switched to T2:

T2_Value{} = B.load(); ---> T2_Value{b} (b++) a=3 b=2
// b is held by B and T2_Value

A.exchange(T2_Value):
{
	T2_Value.incref(); ---> (b++) a=3 b=3
	
	T2_Result = A._Base::exchange(T2_Value); ---> A{b} T2_Result{a}
	// a is held by T2_Result and T1_Value
	// b is held by A, B and T2_Value
}

destroy T2_Value; ---> (b--) a=3 b=2
// b is held by A and B

~~~~thread switched to T1:

	T1_Result = B._Base::exchange(T1_Value); ---> B{a} T1_Result{b}
	// a is held by B, T2_Result and T1_Value
	// b is held by A and T1_Result
}

destroy T1_Value; ---> (a--) a=2 b=2
// a is held by B and T2_Result

A.store(T1_Result):
{
	T1_Old = A.exchange(T1_Result):
	{
		T1_Result.incref(); ---> (b++) a=2 b=3
		
		T1_Old = A._Base::exchange(T1_Result); ---> A{b} (no change in A) T1_Old{b}
		// b is held by A, T1_Result and T1_Old
	}
	destroy T1_Old; ---> (b--) a=2 b=2
	// b is held by A and T1_Result
}

destroy T1_Result; ---> (b--) a=2 b=1
// b is held by A

~~~~thread swithed to T2:
// remember that a is still held by B and T2_Result

B.store(T2_Result):
{
	T2_Old = B.exchange(T2_Result):
	{
		T2_Result.incref(); ---> (a++) a=3 b=1
		
		T2_Old = B._Base::exchange(T2_Result); ---> B{a} (no change in B) T2_Old{a}
		// a is held by B, T2_Result and T2_Old
	}
	destroy T2_Old; ---> (a--) a=2 b=1
	// a is held by B and T2_Result
}

destroy T2_Result; ---> (a--) a=1 b=1
// a is held by B

finaly A holds b, B holds a, a=1, b=1 - everything went well
```

### SCENARIO 3 - same as SCENARIO 2 but thread is switched before store()

```
// at the beginning a=1 and b=1
~~~~start at thread T1:

T1_Value{} = A.load(); ---> T1_Value{a} (a++) a=2 b=1
// a is held by A and T1_Value

B.exchange(T1_Value):
{
	T1_Value.incref(); ---> (a++) a=3 b=1

~~~~thread switched to T2:

T2_Value{} = B.load(); ---> T2_Value{b} (b++) a=3 b=2
// b is held by B and T2_Value

A.exchange(T2_Value):
{
	T2_Value.incref(); ---> (b++) a=3 b=3
	
	T2_Result = A._Base::exchange(T2_Value); ---> A{b} T2_Result{a}
	// a is held by T2_Result and T1_Value
	// b is held by A, B and T2_Value
}

destroy T2_Value; ---> (b--) a=3 b=2
// b is held by A and B

~~~~thread switched to T1:

	T1_Result = B._Base::exchange(T1_Value); ---> B{a} T1_Result{b}
	// a is held by B, T2_Result and T1_Value
	// b is held by A and T1_Result
}

destroy T1_Value; ---> (a--) a=2 b=2
// a is held by B and T2_Result

~~~~thread switched to T2:

B.store(T2_Result):
{
	T2_Old = B.exchange(T2_Result);
	{
		T2_Result.incref(); ---> (a++) a=3 b=2
		
		T2_Old = B._Base::exchange(T2_Result); ---> B{a} (no change in B) T2_Old{a}
		// a is held by B, T2_Result and T2_Old
	}
	destroy T2_Old; ---> (a--) a=2 b=2
	// a is held by B and T2_Result
}

destroy T2_Result; ---> (a--) a=1 b=2
// a is held by B

~~~~thread switched to T1:
// remember that b is still held by A and T1_Result

A.store(T1_Result):
{
	T1_Old = A.exchange(T1_Result):
	{
		T1_Result.incref(); ---> (b++) a=1 b=3
		
		T1_Old = A._Base::exchange(T1_Result); ---> A{b} (no change in A) T1_Old{b}
		// b is held by A, T1_Result and T1_Old
	}
	destroy T1_Old; ---> (b--) a=1 b=2
	// b is held by A and T1_Result
}

destroy T1_Result; ---> (b--) a=1 b=1
// b is held by A

finally A holds b, B holds a, a=1, b=1 - everything went well
```